### PR TITLE
Introduce substituted holidays [PoC]

### PR DESCRIPTION
**Pull request by KJhellico**
## Proposed change

Some countries have a well-established practice of moving the working day between public holiday and weekend to another (non-working) day to create a long continuous holidays. For example, in Ukraine in 2021, Independence Day, 24.08, fell on a Tuesday, so the working day from Monday 23.08 was moved to Saturday 28.08, and Monday became a day off.

To support such cases by Python Holidays, I propose the following structure:
```python
    substituted_holidays = {
        ...
        2021: (
            (JAN, 16, JAN, 8),
            (AUG, 28, AUG, 23),
            (OCT, 23, OCT, 15),
        ),
}
```
when 1st pair (AUG, 28) is source non-working date ("date from"), 2nd (AUG, 23) - is target date ("date to"), that becomes a day off.

Substitution from other year is also supported now:
```python
    substituted_holidays = {
        2001: (
            (APR, 28, APR, 30),
            (MAY, 5, MAY, 10),
            (MAY, 6, MAY, 11),
            (JUN, 23, JUN, 29),
            (2001, DEC, 29, DEC, 31),
        ),
    ...
```
All comments and suggestions on naming, format and implementation are highly welcome (starting with name "substituted holidays" itself) :)

## Type of change

- [ ] New country/market holidays support (thank you!)
- [ ] Supported country/market holidays update (calendar discrepancy fix, localization)
- [ ] Existing code/documentation/test/process quality improvement (best practice, cleanup, refactoring, optimization)
- [ ] Dependency update (version deprecation/upgrade)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Breaking change (a code change causing existing functionality to break)
- [x] New feature (new python-holidays functionality in general)

## Checklist

- [x] I've followed the [contributing guidelines][contributing-guidelines]
- [x] This PR is filed against `beta` branch of the repository
- [x] This PR doesn't contain any merge conflicts and has clean commit history
- [x] The code style looks good: `make pre-commit`
- [x] All tests pass locally: `make test`, `make tox` (we strongly encourage adding tests to your code)
- [x] The related [documentation][docs] has been added/updated (check off the box for free if no updates is required)

[contributing-guidelines]: https://github.com/dr-prodigy/python-holidays/blob/beta/CONTRIBUTING.rst
[docs]: https://github.com/dr-prodigy/python-holidays/tree/beta/docs/source

### CODE REVIEW

File: holidays/locale/ar/LC_MESSAGES/UA.po [hunk 1]
```diff
@@ -3,95 +3,101 @@
...
 "Plural-Forms: nplurals=6; plural=(n==0 ? 0 : n==1 ? 1 : n==2 ? 2 : n%100>=3 && n%100<=10 ? 3 : n%100>=11 && n%100<=99 ? 4 : 5);\n"
 "Generated-By: Lingua 4.15.0\n"
-"X-Generator: Poedit 3.3.1\n"
+"X-Generator: Poedit 3.2.2\n"
+
+#. Day off (substituted from {day}.{month}).
+#: ./holidays/countries/ukraine.py:74
+#, python-format
+msgid "Вихідний день (перенесено з {day:02}.{month:02})"
+msgstr "يوم عطلة (استبدل من {day:02}/{month:02})"
```
Comment (KJhellico) [inline holidays/locale/ar/LC_MESSAGES/UA.po:23]: @abh31000 I am not sure if this translation and the chosen date format are correct, so I would be happy to hear your opinion.

File: tests/countries/test_ukraine.py [hunk 2]
```diff
@@ -316,97 +416,83 @@ def test_2022(self):
...
+            ("2021-12-27", "Різдво Христове (за григоріанським календарем) (вихідний)"),
         )
 
     def test_l10n_en_us(self):
         self.assertLocalizedHolidays(
             "en_US",
             ("2021-01-01", "New Year's Day"),
             ("2021-01-07", "Christmas (Julian calendar)"),
+            ("2021-01-08", "Day off (substituted from 1/16)"),
             ("2021-03-08", "International Women's Day"),
```
Comment (arkid15r) [inline tests/countries/test_ukraine.py:455]: Not sure if it's valid w/o a year.

**Review by arkid15r — COMMENTED** [review]
The base idea implementation looks great! I suggested a couple of aspects to think through:

File: holidays/holiday_base.py [hunk 3]
```diff
@@ -557,6 +560,18 @@ def _add_subdiv_holidays(self):
             if add_subdiv_holidays and callable(add_subdiv_holidays):
                 add_subdiv_holidays()
 
+    def _add_substituted_holidays(self):
+        """Populate substituted holidays."""
+        substituted_label = getattr(
```
[the remark this answers is not in the document]
Comment (KJhellico) [reply holidays/holiday_base.py:565]: Hmm, that makes sense. But which exception is more applicable here?

File: holidays/locale/ar/LC_MESSAGES/UA.po (hunk 1 above)
Comment (abh31000) [reply holidays/locale/ar/LC_MESSAGES/UA.po:22]:
```suggestion
msgstr "يوم عطلة (استبدل من {يوم:02}.{شهر:02})"
```
Yeah like this it's correct. I also think "day" and "month" should be translated

Comment (KJhellico) [reply holidays/locale/ar/LC_MESSAGES/UA.po:22]: No, here "day" and "month" is variable substitutions for` str.format()`:
```python
self._add_holiday(
    self.tr(substituted_label).format(day=from_day, month=from_month), to_month, to_day
)
```
And it looks like I'm going to replace it. :)

Comment (abh31000) [reply holidays/locale/ar/LC_MESSAGES/UA.po:22]: Ah my bad ! :sleeping: 
Both the translation and date format are correct.

Comment (KJhellico) [reply holidays/locale/ar/LC_MESSAGES/UA.po]: Now date format and substituted holiday name is set separately.

File: holidays/holiday_base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (arkid15r) [reply holidays/holiday_base.py]: It could be just a ValueError until we migrate to custom exceptions approach.

**Review by arkid15r — COMMENTED** [review]
That's a PoC we could start using 👍

**Pull request merged**